### PR TITLE
SAR Penetration Testing Findings

### LINKED ISSUE #664: SAR: Penetration Testing-Findings - 4.8
**Extended Description**
- As an ASAP Developer, in order to provide FedRAMP reviewers easy confirmation that penetration testing findings have been properly identified, I want to address items regarding penetration testing.

**Acceptance Criteria**
- [x] All Schematron assertion messages are declarative statements which affirm the positive test outcome.
- [x] All Schematron assertion diagnostic messages are declarative statements which explain the negative test outcome.
- [x] The Schematron code has no assertion failures when validated using `src/validations/styleguides/sch.sch` using the `basic` phase.
- [x] XSpec unit tests for positive and negative Schematron assertion outcomes accompany all Schematron assertions (where feasible).
- [x] An observation with a method='TEST' and type='finding' has subject[@type='component'] whose relevant-evidence/@href matches the uuid of the penetration-test-output resource in the back-matter.
- [x] An observation as above has an origin/actor with an actor-uuid that matches a responsible-party of the role-id of 'penetration-test-lead' or 'penetration-test-team'
- [x] The count of responsible-party[@role-id='penetration-test-lead']/party-uuid is one.

**Definition of Done**
 - [x] Acceptance criteria met
 - [x] Unit test coverage of our code > 95%
 - [x] Automated code quality checks passed
 - [x] Security reviewed and reported
 - [x] Reviewed against plain language guidelines
 - [x] Code must be self-documenting
 - [x] No local tech debt
 - [x] Load/performance tests passed – needs to be created/automated
 - [x] Documentation updated
 - [x] Architectural Decision Record completed as necessary for significant design choices
 - [ ] PR reviewed & approved
 - [ ] Source code merged

### PR DESCRIPTION
**Pull request by markXLIX**
Closes #664 
* An observation with a method='TEST' and type='finding' has subject[@type='component'] whose relevant-evidence/@href matches the uuid of the penetration-test-output resource in the back-matter.
* An observation as above has an origin/actor with an actor-uuid that matches a responsible-party of the role-id of 'penetration-test-lead' or 'penetration-test-team'
* The count of responsible-party[@role-id='penetration-test-lead']/party-uuid is one.

### CODE REVIEW

File: src/validations/rules/sar.sch [hunk 1]
```diff
@@ -176,6 +176,53 @@
...
+                    if (oscal:type = 'finding' and oscal:method = 'TEST' and not(oscal:subject/@type = 'component'))
+                    then
+                        if (oscal:relevant-evidence[substring-after(@href, '#') = /oscal:assessment-results/oscal:back-matter/oscal:resource[
+                        oscal:prop[@value = 'penetration-test-report']]/@uuid])
+                        then
+                            true()
+                        else
+                            false()
+                    else
+                        true()">An observation has a relevant-evidence href that matches the penetration-test-report back-matter
```
Comment (danielnaab) [inline src/validations/rules/sar.sch:199]:
```suggestion
                        true()">An observation must have a relevant-evidence href that matches the penetration-test-report back-matter
```

File: src/validations/rules/sar.sch [hunk 2]
```diff
@@ -176,6 +176,53 @@
...
+                    if (oscal:type = 'finding' and oscal:method = 'TEST' and not(oscal:subject/@type = 'component') and oscal:relevant-evidence[substring-after(@href, '#') = /oscal:assessment-results/oscal:back-matter/oscal:resource[
+                    oscal:prop[@value = 'penetration-test-report']]/@uuid])
+                    then
+                        if (oscal:origin/oscal:actor[@type = 'party']/@actor-uuid[. = $pen-test-team])
+                        then
+                            true()
+                        else
+                            false()
+                    else
+                        true()">A penetration test observation has actors that are described in the responsible-party assemblies for
```
Comment (danielnaab) [inline src/validations/rules/sar.sch:221]:
```suggestion
                        true()">A penetration test observation must have actors that are described in the responsible-party assemblies for
```

File: src/validations/rules/sar.sch [hunk 3]
```diff
@@ -258,6 +305,20 @@
...
+        <sch:rule
+            context="oscal:metadata">
+            <sch:assert
+                diagnostics="has-pen-test-lead-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP Security Assessment Plans (SAR) §4.8"
+                fedramp:specific="true"
+                id="has-pen-test-lead"
+                role="error"
+                test="exists(oscal:responsible-party[@role-id = 'penetration-test-lead']/oscal:party-uuid) and not(exists(oscal:responsible-party[@role-id = 'penetration-test-lead']/oscal:party-uuid[2]))">The
+                count of party-uuid elements of a responsible-party of role-id 'penetration-test-lead' is one.</sch:assert>
```
Comment (danielnaab) [inline src/validations/rules/sar.sch:319]:
```suggestion
                count of party-uuid elements of a responsible-party of role-id 'penetration-test-lead' must be one.</sch:assert>
```

File: src/validations/test/rules/sar.xspec [hunk 4]
```diff
@@ -296,6 +296,206 @@
 
         </x:scenario>
 
+        <x:scenario
+            label="when the metadata">
+            <x:scenario
+                label="has a responsible-party of role-id 'penetration-test-lead">
```
Comment (danielnaab) [inline src/validations/test/rules/sar.xspec:302]:
```suggestion
                label="has a responsible-party of role-id 'penetration-test-lead'">
```

File: src/validations/test/rules/sar.xspec [hunk 5]
```diff
@@ -296,6 +296,206 @@
...
+                                    role-id="penetration-test-lead">
+                                    <party-uuid>e934d8b5-13e5-4f77-b55e-871e6f2df2fe</party-uuid>
+                                    <party-uuid>e934d8b5-13e5-4f77-b55e-871e6f2df2ff</party-uuid>
+                                </responsible-party>
+                            </metadata>
+                        </assessment-results>
+                    </x:context>
+                    <x:expect-assert
+                        id="has-pen-test-lead"
+                        label="this is incorrect" />
```
Comment (danielnaab) [inline src/validations/test/rules/sar.xspec:336]: this is in a few places, but I think most of the asserts use this language. It reads fine either was, so 🤷 

```suggestion
                        label="that is incorrect" />
```

File: src/validations/test/rules/sar.xspec [hunk 6]
```diff
@@ -296,6 +296,206 @@
...
+            <x:scenario
+                label="when a result">
+                <x:scenario
+                    label="has an observation">
+                    <x:scenario
+                        label="with a type of 'finding' and a method of 'TEST'">
+                        <x:scenario
+                            label="and does not have a subject of type component">
+                            <x:scenario
+                                label="and whose relevant-evidence/@href value matches a the penetration-test-report resource in the back-matter">
```
Comment (danielnaab) [inline src/validations/test/rules/sar.xspec:352]:
```suggestion
                                label="and whose relevant-evidence/@href value matches the penetration-test-report resource in the back-matter">
```

File: src/validations/test/rules/sar.xspec [hunk 7]
```diff
@@ -296,6 +296,206 @@
...
+                                                </resource>
+                                            </back-matter>
+                                        </assessment-results>
+                                    </x:context>
+                                    <x:expect-not-assert
+                                        id="has-pen-test-team-match"
+                                        label="this is correct" />
+                                </x:scenario>
+                                <x:scenario
+                                    label="and has party that does not matche the penetration-test-lead responsible-party">
```
Comment (danielnaab) [inline src/validations/test/rules/sar.xspec:426]:
```suggestion
                                    label="and has party that does not match the penetration-test-lead responsible-party">
```

File: src/validations/test/rules/sar.xspec [hunk 8]
```diff
@@ -296,6 +296,206 @@
...
+                                            </back-matter>
+                                        </assessment-results>
+                                    </x:context>
+                                    <x:expect-assert
+                                        id="has-pen-test-team-match"
+                                        label="this is incorrect" />
+                                </x:scenario>
+                            </x:scenario>
+                            <x:scenario
+                                label="and whose relevant-evidence/@href value does not match a the penetration-test-report resource in the back-matter">
```
Comment (danielnaab) [inline src/validations/test/rules/sar.xspec:466]:
```suggestion
                                label="and whose relevant-evidence/@href value does not match the penetration-test-report resource in the back-matter">
```

**Review by danielnaab — CHANGES_REQUESTED** [review]
wording changes

File: src/validations/test/rules/sar.xspec (hunk 5 above)
Comment (markXLIX) [reply src/validations/test/rules/sar.xspec:336]: I feel like, if this/that is desired we should just do a global SR on the xpsecs and commit that as a single story.  I'll start an issue.

**Review by danielnaab — APPROVED** [review]

File: src/validations/rules/sar.sch [hunk 9]
```diff
@@ -176,6 +176,53 @@
...
+        <sch:rule
+            context="oscal:observation">
+            <sch:assert
+                diagnostics="has-pen-test-finding-resource-diagnostic"
+                doc:guide-reference="Guide to OSCAL-based FedRAMP Security Assessment Plans (SAR) §4.8"
+                fedramp:specific="true"
+                id="has-pen-test-finding-resource"
+                role="error"
+                test="
+                    if (oscal:type = 'finding' and oscal:method = 'TEST' and not(oscal:subject/@type = 'component'))
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sar.sch:190]: No reason I can see.  I must have left in a debug item.  When testing now without that not() three errors are found but they are correctly errors as the resource matching does not have a prop with @value = 'penetration-test-report'.

This is corrected.

**Pull request merged**